src/gossip.rs

```diff
@@ -0,0 +1,90 @@
+use crate::logger::{log_error, log_info, FilesystemLogger, Logger};
+use crate::types::{GossipSync, NetworkGraph, P2PGossipSync, RapidGossipSync};
+use crate::Error;
+
+use lightning::routing::utxo::UtxoLookup;
+
+use std::sync::atomic::{AtomicU32, Ordering};
+use std::sync::Arc;
+
+pub(crate) enum GossipSource {
+	P2PNetwork {
+		gossip_sync: Arc<P2PGossipSync>,
+	},
+	RapidGossipSync {
+		gossip_sync: Arc<RapidGossipSync>,
+		server_url: String,
+		latest_sync_timestamp: AtomicU32,
+		logger: Arc<FilesystemLogger>,
+	},
+}
+
+impl GossipSource {
+	pub fn new_p2p(network_graph: Arc<NetworkGraph>, logger: Arc<FilesystemLogger>) -> Self {
+		let gossip_sync = Arc::new(P2PGossipSync::new(
+			network_graph,
+			None::<Arc<dyn UtxoLookup + Send + Sync>>,
+			logger,
+		));
+		Self::P2PNetwork { gossip_sync }
+	}
+
+	pub fn new_rgs(
+		server_url: String, network_graph: Arc<NetworkGraph>, logger: Arc<FilesystemLogger>,
+	) -> Self {
+		let gossip_sync = Arc::new(RapidGossipSync::new(network_graph, Arc::clone(&logger)));
+		let latest_sync_timestamp = AtomicU32::new(0);
+		Self::RapidGossipSync { gossip_sync, server_url, latest_sync_timestamp, logger }
+	}
+
+	pub fn is_p2p(&self) -> bool {
+		if let Self::P2PNetwork { .. } = self {
+			true
+		} else {
+			false
+		}
+	}
+
+	pub fn is_rgs(&self) -> bool {
+		!self.is_p2p()
```

Comment:
May want to explicitly check for `RapidGossipSync` in case the enum changes in the future to have more variants.